designs/2018-programmatic-style-edits/README.md

```diff
@@ -0,0 +1,401 @@
+1. Allow turning on/off style editors using `--fix-type`
+1. Ensure style editors are composable
+
+Definitions:
+
+* **Style Edit** - an individual code modification written in JavaScript.
+* **Style Editor** - a collection of style edits to apply in order. Style editors are applied through configuration.
+
+### Style Editors in Plugins
+
+A plugin can contain one or more style editors, exported on a `styles` object in a manner similar to `rules` and `processors`, for example:
+
+```js
+// eslint-plugin-mystyles
+const es5StyleEditor = require("./styles/es5");
+
+module.exports = {
+    styles: {
+        es5: es5StyleEditor
+    }
+};
+```
+
+In this code, the plugin `eslint-plugin-mystyles` exports a style editor called `es5`. In order to apply the style editor to code, the end-user would add a `style` property to the configuration object:
+
+```js
+module.exports = {
+    plugins: ["mystyles"],
+    parserOptions: {
+        ecmaVersion: 5
+    },
+    style: "mystyles/es5"
+}
+```
+
+The `style` property can be used at the top level of the configuration to apply to all files or in an `overrides` object to apply to a subset of files.
+
+Options for style editors can be specified using a `styleOptions` key, as in this example:
+
+```js
+module.exports = {
+    plugins: ["mystyles"],
+    parserOptions: {
+        ecmaVersion: 5
+    },
+    style: "mystyles/es5",
+    styleOptions: {
+        indent: "tabs"
+    }
+}
+```
+
+The `styleOptions` property is just an opaque object to ESLint, without any predefined properties. Style editors can define their options and ESLint will pass the `styleOptions` object directly into the style editor through `context.options`.
+
+### Style Editor Format
+
+A style editor is an object literal with a `meta` property to provide information about the style editor and an `edits` property that is an array of the style edits to apply, for example:
+
+```js
+// style editor format
+module.exports = {
+    meta: {
+        name: "ES5 Style",
+        description: "A style editor just for ES5"
+    },
+    edits: [
+        // individual style edits
+    ]
+}
+```
+
+The `edits` array contains individual objects in this form:
+
+```js
+{
+    type: "edit type",
+    edit(context) {
+        // modify the source code
+    }
+}
+```
+
+This format of style edit is intentionally generic so that we can add new style edit types in the future. For the purposes of this proposal, I am starting with just one type: `text`. An example text edit looks like this:
+
+```js
+{
+    type: "text",
+    edit(context) {
+        return context.sourceText.replace(/\r\n?/g, "\n");
+    }
+}
+```
+
+This fixer simply normalizes line endings from Windows-style to Unix-style. It accesses the source code text through `context.sourceText` and returns a string. Each `text` fixer must return a string with the modifications to the source code.
+
+The `context` object has the following properties:
```

Comment:
Unless the user is using some 3rd party tool to parse and format the code, it seems like it would be fairly difficult to write a formatter. Do you see the main use-case here as using something like Prettier or Recast in the style enforcer?